# Data Cleaning

## 1. Loading the Data

In [564]:
import pandas as pd
import numpy as np

In [565]:
forest_area = pd.read_csv("dataset/Forest Area.csv")
freshwater = pd.read_csv("dataset/Freshwater.csv", skiprows=3)
gdp_per_capita = pd.read_csv("dataset/gdp_per_capita.csv")
governance = pd.read_csv("dataset/Governance.csv")
human_development_index = pd.read_csv("dataset/Human Development Index.csv")
sustainable_energy = pd.read_csv("dataset/Sustainable energy for all.csv", skiprows=3)
emissions = pd.read_csv("dataset/trends_in_greenhouse_gas_emissions.csv", skiprows=3)

## 2. Choosing Relevant Variables

### Forest Area - 2020 Forest Area

In [566]:
# Forest area
forest_area_filtered = forest_area[['Country and Area', 'Forest Area, 2020 (1000 ha)']]
forest_area_filtered = forest_area_filtered.rename(columns={'Country and Area': 'Country'})
forest_area_filtered.head()

Country  Forest Area, 2020 (1000 ha)
0           WORLD                   4058930.81
1     Afghanistan                      1208.44
2         Albania                       788.90
3         Algeria                      1949.00
4  American Samoa                        17.13

### Freshwater

In [567]:
freshwater_filtered = freshwater[['Country', 'Internal renewable freshwater resources flows 2020']]
freshwater_filtered.head()

Country Internal renewable freshwater resources flows 2020
0     Afghanistan                                               47.2
1         Albania                                               26.9
2         Algeria                                               11.2
3  American Samoa                                                 ..
4         Andorra                                                0.3

### GDP per Capita

In [568]:
gdp_per_capita_filtered = gdp_per_capita[['Country Name', '2021 [YR2021]']]
gdp_per_capita_filtered = gdp_per_capita_filtered.rename(columns={'Country Name': 'Country'})
gdp_per_capita_filtered = gdp_per_capita_filtered.rename(columns={'2021 [YR2021]': 'GDP per Capita 2021'})
gdp_per_capita_filtered.head()

Country GDP per Capita 2021
0     Afghanistan    355.777826392648
1         Albania    6377.20309553753
2         Algeria    3700.31469728198
3  American Samoa    16653.7137781725
4         Andorra    42072.3194231234

### Governance

In [569]:
governance_filtered = governance[['Country and area', 'Paris Agreement', 'UN Framework Convention on Climate Change']]
governance_filtered = governance_filtered.rename(columns={'Country and area': 'Country'})
governance_filtered.head()

Country Paris Agreement  UN Framework Convention on Climate Change
0  Afghanistan            2017                                       2002
1      Albania            2016                                       1994
2      Algeria            2016                                       1993
3      Andorra            2017                                       2011
4       Angola             ...                                       2000

### Human Development Index

In [570]:
human_development_index_filtered = human_development_index[['Country', 'Human Development Index (2021)']]
human_development_index_filtered.head()

Country  Human Development Index (2021)
0           Afghanistan                           0.478
1                Angola                           0.586
2               Albania                           0.796
3               Andorra                           0.858
4  United Arab Emirates                           0.911

### Sustainable Energy

In [571]:
sustainable_energy_filtered = sustainable_energy[['Country', 'Access to clean fuels and technologies for cooking 2021', 'Renewable energy consumption 2020']]
sustainable_energy_filtered.head()

Country Access to clean fuels and technologies for cooking 2021  \
0     Afghanistan                                               35.4        
1         Albania                                               83.7        
2         Algeria                                               99.7        
3  American Samoa                                                 ..        
4         Andorra                                                100        

  Renewable energy consumption 2020  
0                              17.6  
1                              44.6  
2                               0.2  
3                               0.5  
4                              21.9

### Emissions

In [572]:
emissions_filtered = emissions[['Country', 'Carbon dioxide emissions 2020', 'Methane emissions 2020', 'Nitrous oxide emissions 2020']]
emissions_filtered.head()

Country Carbon dioxide emissions 2020 Methane emissions 2020  \
0     Afghanistan                       8709.47            16222.03596   
1         Albania                        4383.2            2692.195886   
2         Algeria                        161563            86543.92362   
3  American Samoa                            ..                     ..   
4         Andorra                    448.884399              53.600959   

  Nitrous oxide emissions 2020  
0                  4863.386801  
1                  1015.186729  
2                   12578.7476  
3                           ..  
4                     1.758811

## 3. Merge Datasets

Remove unnecessary rows after last country (Zimbabwe) for Freshwater, Sustainable Energy, and Emissions dataset

In [573]:
freshwater_filtered = freshwater_filtered.loc[:freshwater_filtered[freshwater_filtered['Country'] == 'Zimbabwe'].index[0]]
freshwater_filtered

Country Internal renewable freshwater resources flows 2020
0              Afghanistan                                               47.2
1                  Albania                                               26.9
2                  Algeria                                               11.2
3           American Samoa                                                 ..
4                  Andorra                                                0.3
..                     ...                                                ...
209  Virgin Islands (U.S.)                                                 ..
210     West Bank and Gaza                                                0.8
211            Yemen, Rep.                                                2.1
212                 Zambia                                               80.2
213               Zimbabwe                                               12.3

[214 rows x 2 columns]

In [574]:
sustainable_energy_filtered = sustainable_energy_filtered.loc[:sustainable_energy_filtered[sustainable_energy_filtered['Country'] == 'Zimbabwe'].index[0]]
sustainable_energy_filtered

Country  \
0              Afghanistan   
1                  Albania   
2                  Algeria   
3           American Samoa   
4                  Andorra   
..                     ...   
209  Virgin Islands (U.S.)   
210     West Bank and Gaza   
211            Yemen, Rep.   
212                 Zambia   
213               Zimbabwe   

    Access to clean fuels and technologies for cooking 2021  \
0                                                 35.4        
1                                                 83.7        
2                                                 99.7        
3                                                   ..        
4                                                  100        
..                                                 ...        
209                                                 ..        
210                                                 ..        
211                                               61.3        
212                                               10.2        
213                                               30.3        

    Renewable energy consumption 2020  
0                                17.6  
1                                44.6  
2                                 0.2  
3                                 0.5  
4                                21.9  
..                                ...  
209                               5.1  
210                                15  
211                               3.5  
212                              81.8  
213                              84.4  

[214 rows x 3 columns]

In [575]:
emissions_filtered = emissions_filtered.loc[:emissions_filtered[emissions_filtered['Country'] == 'Zimbabwe'].index[0]]
emissions_filtered

Country Carbon dioxide emissions 2020  \
0              Afghanistan                       8709.47   
1                  Albania                        4383.2   
2                  Algeria                        161563   
3           American Samoa                            ..   
4                  Andorra                    448.884399   
..                     ...                           ...   
209  Virgin Islands (U.S.)                            ..   
210     West Bank and Gaza                            ..   
211            Yemen, Rep.                        9960.1   
212                 Zambia                        7607.1   
213               Zimbabwe                        8312.5   

    Methane emissions 2020 Nitrous oxide emissions 2020  
0              16222.03596                  4863.386801  
1              2692.195886                  1015.186729  
2              86543.92362                   12578.7476  
3                       ..                           ..  
4                53.600959                     1.758811  
..                     ...                          ...  
209                     ..                           ..  
210                     ..                           ..  
211            10542.15376                  3364.858401  
212            15448.75233                  12363.49464  
213            12060.82302                  4690.849521  

[214 rows x 4 columns]

Get standardised country names map to merge data into one CSV file as each datasets might have different conventions on naming the countries. 

I did this with the help of GPT-4, but it was not entirely accurate so I also had to input some of the key-value pairs manually.

In [576]:
country_name_mapping = {
    "Bahamas, The": "Bahamas",
    "Bolivia": "Bolivia (Plurinational State of)",
    "Brunei": "Brunei Darussalam",
    "Congo, Dem. Rep.": "Democratic Republic of the Congo",
    "Congo, Democratic Republic of the": "Democratic Republic of the Congo",
    "The Democratic Republic of the Congo": "Democratic Republic of the Congo",
    "Congo, Rep.": "Congo",
    "Congo, Republic of": "Congo",
    "Cote d'Ivoire": "Côte d’Ivoire",
    "Ivory Coast": "Côte d’Ivoire",
    "Czech Republic": "Czechia",
    "Gambia, The": "Gambia",
    "Korea, Rep.": "Republic of Korea",
    "Korea, Republic of": "Republic of Korea",
    "South Korea": "Republic of Korea",
    "Korea, Dem. People's Rep.": "Democratic People's Republic of Korea",
    "North Korea": "Democratic People's Republic of Korea",
    "Iran, Islamic Rep.": "Iran (Islamic Republic of)",
    "Iran": "Iran (Islamic Republic of)",
    "Lao PDR": "Lao People's Democratic Republic",
    "Lao": "Lao People's Democratic Republic",
    "Micronesia, Fed. Sts.": "Micronesia (Federated States of)",
    "Micronesia": "Micronesia (Federated States of)",
    "Moldova": "Republic of Moldova",
    "Palestine, State of": "State of Palestine",
    "St. Kitts and Nevis": "Saint Kitts and Nevis",
    "St. Lucia": "Saint Lucia",
    "St. Vincent and the Grenadines": "Saint Vincent and the Grenadines",
    "Turkiye": "Turkey",
    "Tanzania": "United Republic of Tanzania",
    "Venezuela, RB": "Venezuela",
    "Yemen, Rep.": "Yemen",
    "Syria": "Syrian Arab Republic",
    "Egypt, Arab Rep.": "Egypt",
    "Kyrgyz Republic": "Kyrgyzstan",
    "Slovak Republic": "Slovakia",
    "Russia": "Russian Federation",
    "United Kingdom": "United Kingdom of Great Britain and Northern Ireland",
    "United States": "United States of America",
    "Vietnam": "Viet Nam",
    "Venezuela": "Venezuela (Bolivarian Republic of)",
    "Venezuela, RB": "Venezuela (Bolivarian Republic of)",
    "West Bank and Gaza": "State of Palestine",
}

In [577]:
print("Forest Area: ", len(forest_area_filtered))
print("Freshwater: ", len(freshwater_filtered))
print("GDP per Capita: ", len(gdp_per_capita_filtered))
print("Governance: ", len(governance_filtered))
print("HDI: ", len(human_development_index_filtered))
print("Sustainable Energy: ", len(sustainable_energy_filtered))
print("Emissions: ", len(emissions_filtered))

Forest Area:  237
Freshwater:  214
GDP per Capita:  271
Governance:  194
HDI:  195
Sustainable Energy:  214
Emissions:  214


As the Governance dataset has the least amount of countries, I'm going to use countries from the Governance dataset as the base country list for the merged data.

In [578]:
countries_list = governance_filtered['Country'].unique()
countries_list

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia (Plurinational State of)', 'Bosnia and Herzegovina',
       'Botswana', 'Brazil', 'Brunei Darussalam', 'Bulgaria',
       'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cambodia', 'Cameroon',
       'Canada', 'Central African Republic', 'Chad', 'Chile', 'China',
       'Colombia', 'Comoros', 'Congo', 'Costa Rica', 'Côte d’Ivoire',
       'Croatia', 'Cuba', 'Cyprus', 'Czechia',
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Denmark', 'Djibouti',
       'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France', 'Gabon',
       'Gambia', 'Georgia', 'G

Map the country names from other datasets to match the country name in the Governance dataset.

In [579]:
forest_area_filtered.loc[:, 'Country'] = forest_area_filtered['Country'].replace(country_name_mapping)
forest_area_filtered = forest_area_filtered[forest_area_filtered['Country'].isin(countries_list)]
len(forest_area_filtered)

194

In [580]:
freshwater_filtered.loc[:, 'Country'] = freshwater_filtered['Country'].replace(country_name_mapping)
freshwater_filtered = freshwater_filtered[freshwater_filtered['Country'].isin(countries_list)]
len(freshwater_filtered)

193

In [581]:
gdp_per_capita_filtered.loc[:, 'Country'] = gdp_per_capita_filtered['Country'].replace(country_name_mapping)
gdp_per_capita_filtered = gdp_per_capita_filtered[gdp_per_capita_filtered['Country'].isin(countries_list)]
len(gdp_per_capita_filtered)

194

In [582]:
human_development_index_filtered.loc[:, 'Country'] = human_development_index_filtered['Country'].replace(country_name_mapping)
human_development_index_filtered = human_development_index_filtered[human_development_index_filtered['Country'].isin(countries_list)]
len(human_development_index_filtered)

194

In [583]:
sustainable_energy_filtered.loc[:, 'Country'] = sustainable_energy_filtered['Country'].replace(country_name_mapping)
sustainable_energy_filtered = sustainable_energy_filtered[sustainable_energy_filtered['Country'].isin(countries_list)]
len(sustainable_energy_filtered)

193

In [584]:
emissions_filtered.loc[:, 'Country'] = emissions_filtered['Country'].replace(country_name_mapping)
emissions_filtered = emissions_filtered[emissions_filtered['Country'].isin(countries_list)]
len(emissions_filtered)

193

In [585]:
freshwater_countries = freshwater_filtered['Country'].unique()
missing_countries = list(set(countries_list) - set(freshwater_countries))
missing_countries

['Nauru']

In [586]:
sustainable_energy_countries = sustainable_energy_filtered['Country'].unique()
missing_countries = list(set(countries_list) - set(sustainable_energy_countries))
missing_countries

['Nauru']

In [587]:
emissions_countries = emissions_filtered['Country'].unique()
missing_countries = list(set(countries_list) - set(emissions_countries))
missing_countries

['Nauru']

So in conclusion, I am going to have a total of 193 countries.

The country excluded from the index that is in the original countries list is:
* Nauru

as it is not present in the Freshwater, Sustainable Energy, and Emissions dataset.

In [588]:
from functools import reduce

In [589]:
dfs = [forest_area_filtered, freshwater_filtered, gdp_per_capita_filtered, governance_filtered, human_development_index_filtered, sustainable_energy_filtered, emissions_filtered]
merge_dfs = lambda left, right: pd.merge(left, right, on='Country', how='inner')
lcdci_df = reduce(merge_dfs, dfs)
lcdci_df

Country  Forest Area, 2020 (1000 ha)  \
0                           Afghanistan                      1208.44   
1                               Albania                       788.90   
2                               Algeria                      1949.00   
3                               Andorra                        16.00   
4                                Angola                     66607.38   
..                                  ...                          ...   
188  Venezuela (Bolivarian Republic of)                     46230.90   
189                            Viet Nam                     14643.09   
190                               Yemen                       549.00   
191                              Zambia                     44814.03   
192                            Zimbabwe                     17444.58   

    Internal renewable freshwater resources flows 2020 GDP per Capita 2021  \
0                                                 47.2    355.777826392648   
1                                                 26.9    6377.20309553753   
2                                                 11.2    3700.31469728198   
3                                                  0.3    42072.3194231234   
4                                                  148    1927.47407832091   
..                                                 ...                 ...   
188                                                805                  ..   
189                                              359.4    3756.48890080884   
190                                                2.1    543.637537619208   
191                                               80.2    1134.71345424991   
192                                               12.3    1773.92041090285   

    Paris Agreement  UN Framework Convention on Climate Change  \
0              2017                                       2002   
1              2016                                       1994   
2              2016                                       1993   
3              2017                                       2011   
4               ...                                       2000   
..              ...                                        ...   
188            2017                                       1994   
189            2016                                       1994   
190             ...                                       1996   
191            2016                                       1993   
192            2017                                       1992   

     Human Development Index (2021)  \
0                             0.478   
1                             0.796   
2                             0.745   
3                             0.858   
4                             0.586   
..                              ...   
188                           0.691   
189                           0.703   
190                           0.455   
191                           0.565   
192                           0.593   

    Access to clean fuels and technologies for cooking 2021  \
0                                                 35.4        
1                                                 83.7        
2                                                 99.7        
3                                                  100        
4                                                   50        
..                                                 ...        
188                                               95.5        
189                                               96.1        
190                                               61.3        
191                                               10.2        
192                                               30.3        

    Renewable energy consumption 2020 Carbon dioxide emissions 2020  \
0                                17.6                       8709.47   
1                                44.6                        4383.2  

In [590]:
# lcdci_df.to_csv("dataset/lcdci.csv", index=False)

In [591]:
lcdci_df.dtypes

Country                                                     object
Forest Area, 2020 (1000 ha)                                float64
Internal renewable freshwater resources flows 2020          object
GDP per Capita 2021                                         object
Paris Agreement                                             object
UN Framework Convention on Climate Change                    int64
Human Development Index (2021)                             float64
Access to clean fuels and technologies for cooking 2021     object
Renewable energy consumption 2020                           object
Carbon dioxide emissions 2020                               object
Methane emissions 2020                                      object
Nitrous oxide emissions 2020                                object
dtype: object

In [594]:
lcdci_df = lcdci_df.rename(columns={'Forest Area, 2020 (1000 ha)': 'forest_area'})
lcdci_df = lcdci_df.rename(columns={'Internal renewable freshwater resources flows 2020': 'freshwater'})
lcdci_df = lcdci_df.rename(columns={'GDP per Capita 2021': 'gdp_per_capita'})
lcdci_df = lcdci_df.rename(columns={'Paris Agreement': 'paris_agreement'})
lcdci_df = lcdci_df.rename(columns={'UN Framework Convention on Climate Change': 'unfccc'})
lcdci_df = lcdci_df.rename(columns={'Human Development Index (2021)': 'hdi'})
lcdci_df = lcdci_df.rename(columns={'Access to clean fuels and technologies for cooking 2021': 'clean_fuel_access'})
lcdci_df = lcdci_df.rename(columns={'Renewable energy consumption 2020': 'renewable_energy_consumption'})
lcdci_df = lcdci_df.rename(columns={'Carbon dioxide emissions 2020': 'co2_emissions'})
lcdci_df = lcdci_df.rename(columns={'Methane emissions 2020': 'methane_emissions'})
lcdci_df = lcdci_df.rename(columns={'Nitrous oxide emissions 2020': 'nitrous_oxide_emissions'})


In [595]:
lcdci_df.head()

Country  forest_area freshwater    gdp_per_capita paris_agreement  \
0  Afghanistan      1208.44       47.2  355.777826392648            2017   
1      Albania       788.90       26.9  6377.20309553753            2016   
2      Algeria      1949.00       11.2  3700.31469728198            2016   
3      Andorra        16.00        0.3  42072.3194231234            2017   
4       Angola     66607.38        148  1927.47407832091             ...   

   unfccc    hdi clean_fuel_access renewable_energy_consumption co2_emissions  \
0    2002  0.478              35.4                         17.6       8709.47   
1    1994  0.796              83.7                         44.6        4383.2   
2    1993  0.745              99.7                          0.2        161563   
3    2011  0.858               100                         21.9    448.884399   
4    2000  0.586                50                           61       19814.5   

  methane_emissions nitrous_oxide_emissions  
0       16222.03596             4863.386801  
1       2692.195886             1015.186729  
2       86543.92362              12578.7476  
3         53.600959                1.758811  
4       32703.23811             17757.86784